File: apps/cdn-node/src/router.rs
use async_nats::jetstream::Message;
use makoto_logger::error;

use mafuyu_nats::{tools::Tools, route::RouteResult, payload::cdn::{subject, PublishImageRequest}};
use mafuyu_nats::errors::NatsHandleError;

pub struct Router {
    cdn_fs: crate::fs::CDNFs
}

impl Router {
    pub async fn new(cdn_fs: crate::fs::CDNFs) -> Self {

        Self {
            cdn_fs
        }
    }

    pub async fn handle(&self, message: Message) {
        if let Err(err) =  message.ack().await {
            error!("[cannot ack] {err}");
            return;
        }

        let subject = match Tools::get_subject(&message.subject) {
            Ok(subject) => subject,
            Err(err) => {
                error!("[subject error]: {err}");
                return;
            }
        };


        let r = match subject.as_str() {
            subject::PUBLISH_IMAGE_PARSED => self.publish_image(message).await,
            _ => {
                error!("[subject not found]");
                return;
            }
        };

        if let Err(err) = r {
            error!("[router error] {err}");
        };

    }

    pub async fn publish_image(&self, message: Message) -> RouteResult {
        let payload = Tools::get_payload::<PublishImageRequest>(&message.payload)?;


        self.cdn_fs.save_image(&payload.filename, payload.base64_image, payload.image_ext).internal_error()?;

        Ok(())
    }
}

File: apps/mailer-async/internal/router.go
package internal

import (
	mafuyuNats "github.com/dehwyy/mafuyu/libs/nats/pkg"
	mafuyuNatsError "github.com/dehwyy/mafuyu/libs/nats/pkg/errors"
	mafuyuNatsPayload "github.com/dehwyy/mafuyu/libs/nats/pkg/payload"
	"github.com/nats-io/nats.go/jetstream"
)

type MailService interface {
	SendEmail(to []string, subject string, content []byte) error
}

type Router struct {
	mailService MailService
}

func NewRouter(mailService MailService) *Router {
	return &Router{
		mailService: mailService,
	}
}

func (r *Router) Handle(msg jetstream.Msg) error {
	if err := msg.Ack(); err != nil {
		return err
	}

	tools := mafuyuNats.NewTools(msg)
	subject, err := tools.GetSubject()
	if err != nil {
		return err
	}

	switch subject {
	case "send_email":
		return r.SendMail(msg)
	default:
		return mafuyuNatsError.NewNatsError("subject not found")
	}
}

func (r *Router) SendMail(msg jetstream.Msg) error {
	payload, err := mafuyuNats.GetPayload[mafuyuNatsPayload.SendEmailPayload](msg.Data())
	if err != nil {
		return err
	}

	return r.mailService.SendEmail([]string{payload.To}, payload.Subject, payload.Content)
}
